File: src/texlint/lsp/config_watch.py
"""Spec 011 — `.jss-lint.toml` watcher for the LSP server.

The server registers a glob watcher for ``.jss-lint.toml`` files
under every workspace folder. When the file changes, this module
re-loads the config and signals "every open document needs to be
re-linted". The actual re-lint dispatch happens in the server.
"""

from __future__ import annotations

import dataclasses
import sys
from collections.abc import Callable, Mapping
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from texlint.api import ToolConfig
from texlint.config import (
    _normalise_ignore_rules,
    _normalise_severity_overrides,
)
from texlint.config import load as load_config

if sys.version_info >= (3, 11):
    import tomllib  # type: ignore[unused-ignore]
else:  # pragma: no cover - covered only on 3.10
    import tomli as tomllib


@dataclass
class ConfigState:
    """Running cache of the active ``.jss-lint.toml`` plus the
    client-pushed (VS Code settings) layer.

    ``path`` is the file we're watching (or ``None`` when no config
    has been picked up yet). ``cfg`` is the most-recent parsed
    config; on a malformed reload we keep the previous ``cfg`` and
    surface a log message via *log*. ``client_settings`` is the raw
    ``jssStyleChecker`` section from the latest
    ``workspace/didChangeConfiguration`` push; ``run_on`` controls
    whether ``didChange`` lints ("change", the default) or only
    ``didSave``/``didOpen`` do ("save").
    """

    path: Path | None = None
    cfg: ToolConfig = None  # type: ignore[assignment]
    last_error: str | None = None
    client_settings: dict[str, Any] = field(default_factory=dict)
    run_on: str = "change"

    def __post_init__(self) -> None:
        if self.cfg is None:
            self.cfg = ToolConfig()

    def effective(self) -> ToolConfig:
        """The config rules actually run with: ``.jss-lint.toml`` as
        the base, client settings layered on top."""
        return merge_client_settings(self.cfg, self.client_settings)


def merge_client_settings(
    cfg: ToolConfig, settings: Mapping[str, Any]
) -> ToolConfig:
    """Layer the client's ``jssStyleChecker`` settings over *cfg*.

    Precedence rules (additive, never wholesale — VS Code pushes its
    defaults for every key on every change, so "client replaces file"
    would silently erase ``.jss-lint.toml`` values):

    * ``ignoreRules`` **unions** into ``ignore_rules``;
    * ``severityOverrides`` **dict-updates** over file overrides
      (client wins per rule id);
    * ``codeWidth`` replaces ``code_width`` (the extension default 80
      equals the tool default, so an unset client is a no-op).

    ``runOn`` is deliberately not a ToolConfig concern — the server
    stores it on :class:`ConfigState` and gates ``didChange`` lints.
    """
    if not settings:
        return cfg
    changes: dict[str, Any] = {}
    if settings.get("ignoreRules"):
        extra = _normalise_ignore_rules(settings["ignoreRules"])
        if extra:
            changes["ignore_rules"] = cfg.ignore_rules | extra
    if settings.get("severityOverrides"):
        client_overrides = _normalise_severity_overrides(
            settings["severityOverrides"]
        )
        if client_overrides:
            changes["severity_overrides"] = {
                **cfg.severity_overrides,
                **client_overrides,
            }
    code_width = settings.get("codeWidth")
    if isinstance(code_width, (int, float)) and int(code_width) > 0:
        changes["code_width"] = int(code_width)
    if not changes:
        return cfg
    return dataclasses.replace(cfg, **changes)


def reload(state: ConfigState, path: Path, log: Callable[[str], None]) -> bool:
    """Reload the config from *path*. Returns ``True`` when the new
    config replaces the old; ``False`` when the file is malformed
    (the previous config stays active)."""
    try:
        # `load_config` accepts a CWD; we want to read THIS file
        # specifically, so we delegate to its private read by
        # constructing a tomllib.loads directly first.
        text = path.read_text(encoding="utf-8")
        tomllib.loads(text)  # raises on malformed TOML
    except (OSError, tomllib.TOMLDecodeError) as exc:
        state.last_error = f"failed to read {path}: {exc}"
        log(state.last_error)
        return False

    try:
        new_cfg = load_config({}, path.parent)
    except Exception as exc:  # pragma: no cover - load_config is robust
        state.last_error = f"failed to load config at {path}: {exc}"
        log(state.last_error)
        return False

    state.path = path
    state.cfg = new_cfg
    state.last_error = None
    return True

File: rust/jsslint-cli/src/lsp_server/config_watch.rs
//! `.jss-lint.toml` + client-settings state for the LSP server. Mirrors
//! `texlint/lsp/config_watch.py`.

use jsslint_core::config::{RawOverrides, ToolConfig};
use jsslint_core::report::Severity;
use serde_json::Value;
use std::collections::HashSet;
use std::path::{Path, PathBuf};

pub struct ConfigState {
    pub path: Option<PathBuf>,
    pub cfg: ToolConfig,
    pub last_error: Option<String>,
    /// Raw `jssStyleChecker` section from the latest
    /// `workspace/didChangeConfiguration` push.
    pub client_settings: Value,
    /// `"change"` (default) or `"save"` — gates whether `didChange`
    /// lints or only `didSave`/`didOpen` do.
    pub run_on: String,
}

impl Default for ConfigState {
    fn default() -> Self {
        Self {
            path: None,
            cfg: ToolConfig::default(),
            last_error: None,
            client_settings: Value::Object(serde_json::Map::new()),
            run_on: "change".to_string(),
        }
    }
}

impl ConfigState {
    /// The config rules actually run with: `.jss-lint.toml` as the
    /// base, client settings layered on top.
    pub fn effective(&self) -> ToolConfig {
        merge_client_settings(&self.cfg, &self.client_settings)
    }
}

fn normalize_ignore_rules_value(v: &Value) -> HashSet<String> {
    match v {
        Value::String(s) => s
            .split(',')
            .map(|p| p.trim().to_string())
            .filter(|p| !p.is_empty())
            .collect(),
        Value::Array(items) => items
            .iter()
            .filter_map(Value::as_str)
            .map(|s| s.trim().to_string())
            .filter(|s| !s.is_empty())
            .collect(),
        _ => HashSet::new(),
    }
}

fn normalize_severity_overrides_value(
    obj: &serde_json::Map<String, Value>,
) -> Vec<(String, Severity)> {
    obj.iter()
        .filter_map(|(k, v)| {
            let sev = Severity::parse(&v.as_str()?.to_lowercase())?;
            Some((k.trim().to_uppercase(), sev))
        })
        .collect()
}

/// Layers the client's `jssStyleChecker` settings over `cfg`.
/// Precedence is additive, never wholesale (VS Code pushes its
/// defaults for every key on every change, so "client replaces file"
/// would silently erase `.jss-lint.toml` values): `ignoreRules`
/// **unions** into `ignore_rules`; `severityOverrides` **dict-updates**
/// over file overrides (client wins per rule id); `codeWidth` replaces
/// `code_width`. Mirrors `config_watch.py::merge_client_settings`.
pub fn merge_client_settings(cfg: &ToolConfig, settings: &Value) -> ToolConfig {
    let mut out = cfg.clone();
    let Some(obj) = settings.as_object() else {
        return out;
    };
    if obj.is_empty() {
        return out;
    }
    if let Some(v) = obj.get("ignoreRules") {
        let extra = normalize_ignore_rules_value(v);
        if !extra.is_empty() {
            out.ignore_rules.extend(extra);
        }
    }
    if let Some(v) = obj.get("severityOverrides").and_then(Value::as_object) {
        let client_overrides = normalize_severity_overrides_value(v);
        if !client_overrides.is_empty() {
            for (rule_id, sev) in client_overrides {
                out.severity_overrides.insert(rule_id, sev);
            }
        }
    }
    if let Some(n) = obj.get("codeWidth").and_then(Value::as_f64) {
        if n > 0.0 {
            out.code_width = n as u32;
        }
    }
    out
}

/// Reloads the config from `path`. Returns `true` when the new config
/// replaces the old; `false` when the file is malformed (the previous
/// config stays active). Mirrors `config_watch.py::reload`.
pub fn reload(state: &mut ConfigState, path: &Path, mut log: impl FnMut(&str)) -> bool {
    let text = match std::fs::read_to_string(path) {
        Ok(t) => t,
        Err(exc) => {
            let msg = format!("failed to read {}: {exc}", path.display());
            log(&msg);
            state.last_error = Some(msg);
            return false;
        }
    };
    if let Err(exc) = text.parse::<toml::Table>() {
        let msg = format!("failed to read {}: {exc}", path.display());
        log(&msg);
        state.last_error = Some(msg);
        return false;
    }

    let parent = path.parent().unwrap_or_else(|| Path::new("."));
    let new_cfg = jsslint_core::config::load(parent, &RawOverrides::default());
    state.path = Some(path.to_path_buf());
    state.cfg = new_cfg;
    state.last_error = None;
    true
}
